### PR TITLE
KAFKA-7858: Replace JoinGroup request/response with automated protocol

### PR DESCRIPTION
**Pull request by abbccdda**
Prioritizing this migration because we have blocking feature from KIP-345 part 1: https://github.com/apache/kafka/pull/6177

Upgrade join group protocols could ease the process of adding group instance id towards JoinGroupResponse.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by abbccdda** [comment]
@cmccabe @hachikuji Could you guys give this diff a quick r? This would unblock the progress on KIP-345, thanks!

**Comment by abbccdda** [comment]
cc @vahidhashemian on this thread since the change is pretty straightforward and requires no background context.

**Comment by abbccdda** [comment]
Retest this please

**Comment by abbccdda** [comment]
Retest this please

**Comment by cmccabe** [comment]
@abbccdda: looks good, thanks!

When processing older responses, I think generation ID needs to default to -1 to match the previous behavior.  This should be changed in `JoinGroupResponse.json`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java [hunk 1]
```diff
@@ -202,7 +203,7 @@ public AbstractCoordinator(LogContext logContext,
      */
     protected abstract Map<String, ByteBuffer> performAssignment(String leaderId,
                                                                  String protocol,
-                                                                 Map<String, ByteBuffer> allMemberMetadata);
+                                                                 List<JoinGroupResponseData.JoinGroupResponseMember> allMemberMetadata);
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:206]: Should this be a `JoinGroupResponseDataSet`?

Comment (abbccdda) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:206]: I don't see this struct anywhere, but I think a list should be fine here?

**Comment by abbccdda** [comment]
Retest this please

**Comment by abbccdda** [comment]
@cmccabe @hachikuji mind taking another look?

**Comment by abbccdda** [comment]
@cmccabe Another look when you get time?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java (hunk 1 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:206]: Sorry, I meant `JoinGroupResponseMemberSet`.  Anyway, a list is fine too-- I don't feel that strongly about it.

**Comment by cmccabe** [comment]
LGTM.  Thanks, @abbccdda .

**Comment by abbccdda** [comment]
Thanks! @cmccabe

**Pull request merged**